activesupport/test/message_encryptor_test.rb

```diff
@@ -132,3 +139,111 @@ def munge(base64_string)
       ::Base64.strict_encode64(bits)
     end
 end
+
+module SharedMessageMetadataTests
+  def setup
+    @secret    = SecureRandom.random_bytes(32)
+    @encryptor = ActiveSupport::MessageEncryptor.new(@secret, encryptor_options)
+
+    super
+  end
+
+  def test_encryption_and_decryption_with_same_purpose
+    message = "a security question's answer"
+    encrypted_message = @encryptor.encrypt_and_sign(message, purpose: "login")
+
+    assert_equal message, @encryptor.decrypt_and_verify(encrypted_message, purpose: "login")
+
+    encrypted_message = @encryptor.encrypt_and_sign(message)
+
+    assert_equal message, @encryptor.decrypt_and_verify(encrypted_message)
+  end
+
+  def test_encryption_and_decryption_with_different_purposes_returns_nil
+    message = "phone number: 9123912310"
+    encrypted_message = @encryptor.encrypt_and_sign(message, purpose: :verification)
+
+    assert_nil @encryptor.decrypt_and_verify(encrypted_message, purpose: "sign up")
+    assert_nil @encryptor.decrypt_and_verify(encrypted_message)
+
+    encrypted_message = @encryptor.encrypt_and_sign(message)
+
+    assert_nil @encryptor.decrypt_and_verify(encrypted_message, purpose: 'sign up')
+  end
+
+  def test_purpose_using_symbols
+    message = "address: #23, main street"
+    encrypted_message = @encryptor.encrypt_and_sign(message, purpose: :checkout)
+
+    assert_equal message, @encryptor.decrypt_and_verify(encrypted_message, purpose: :checkout)
+    assert_equal message, @encryptor.decrypt_and_verify(encrypted_message, purpose: 'checkout')
+  end
+
+  def test_passing_expires_at_sets_expiration_date
+    message = "credit card no: 5012-6784-9087-5678"
```

Comment:
You don't have to get so elaborate with the messages. I'd assign a `@message` ivar in `setup` and use that throughout. I'd probably also make `@message` a hash, so we know the complex scenario encryption scenario works.